### PR TITLE
Added TensorFlow Dataset API example.

### PR DESCRIPTION
**Pull request by jsaporta**
Just another MNIST classification example, this time with TF's Dataset API.

This is my first attempt at contributing to an open-source project, so be nice :)

### CODE REVIEW

File: examples/mnist_dataset_api.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+dataset = Dataset.from_tensor_slices((x_train, y_train))
+dataset = dataset.repeat()
+dataset = dataset.shuffle(buffer_size)
+dataset = dataset.batch(batch_size)
+iterator = dataset.make_one_shot_iterator()
+
+# The next_element graph node provides tuples from the dataset.
+# Its first component is passed to the model's Input layer; the
+# second is passed to the compile method.
+next_element = iterator.get_next()
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: Better to have `inputs, targets = iterator.get_next()`

File: examples/mnist_dataset_api.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+train_model.save_weights(weight_path)
+
+# Clean up the TF session.
+K.clear_session()
+
+# Second session to test loading trained model without tensors.
+x_test = x_test.astype(np.float32)
+x_test = np.expand_dims(x_test, -1)
+
+x_test_inp = layers.Input(shape=(x_test.shape[1:]))
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: Surely it should be `shape=x_test.shape[1:]` in the general case.

File: examples/mnist_dataset_api.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+from tensorflow.contrib.data import Dataset
+
+
+if K.backend() != 'tensorflow':
+    raise RuntimeError('This example can only run with the TensorFlow backend,'
+                       ' because it requires the Datset API, which is not'
+                       ' supported on other platforms.')
+
+
+def cnn_layers(x_train_input):
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: Please rename argument to `inputs`

File: examples/mnist_dataset_api.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    x = layers.MaxPooling2D(pool_size=(2, 2))(x)
+    x = layers.Conv2D(64, (3, 3), activation='relu')(x)
+    x = layers.MaxPooling2D(pool_size=(2, 2))(x)
+    x = layers.Flatten()(x)
+    x = layers.Dense(512, activation='relu')(x)
+    x = layers.Dropout(0.5)(x)
+    x_train_out = layers.Dense(classes,
+                               activation='softmax',
+                               name='x_train_out')(x)
+    return x_train_out
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: Please rename to `predictions` (same for layer name above)

File: examples/mnist_dataset_api.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+batch_size = 128
+buffer_size = 10000
+steps_per_epoch = np.ceil(60_000 / 128).astype('int')  # = 469
+epochs = 5
+classes = 10
+
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+x_train = x_train.astype(np.float32) / 255
+x_train = np.expand_dims(x_train, -1)
+y_train = tf.one_hot(y_train, classes)
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: You don't use the `_train` arrays so you can `_` them altogether.

Also: please put this block when you need it, i.e. at test time (see comment below)

File: examples/mnist_dataset_api.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                steps_per_epoch=steps_per_epoch)
+
+# Save the model weights.
+weight_path = os.path.join(tempfile.gettempdir(), 'saved_wt.h5')
+train_model.save_weights(weight_path)
+
+# Clean up the TF session.
+K.clear_session()
+
+# Second session to test loading trained model without tensors.
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: Please put mnist test data loading/preprocessing here.

File: examples/mnist_dataset_api.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+test_model = keras.models.Model(inputs=x_test_inp, outputs=test_out)
+
+test_model.load_weights(weight_path)
+test_model.compile(optimizer='rmsprop',
+                   loss='categorical_crossentropy',
+                   metrics=['accuracy'])
+test_model.summary()
+
+loss, acc = test_model.evaluate(x_test,
+                                keras.utils.to_categorical(y_test),
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: You could also use `sparse_categorical_crossentropy` and skip `to_categorical`

File: examples/mnist_dataset_api.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+to its get_next() method. Its components are then provided
+to the network's Input layer and the Model.compile() method,
+respectively.
+
+Note that from TensorFlow 1.4, tf.contrib.data is deprecated
+and tf.data is preferred. See the release notes for details.
+
+This example is intended to closely follow the
+mnist_tfrecord.py example. No effort is made for Python 2
+compatibility.
```
Comment (fchollet) [inline examples/mnist_dataset_api.py]: Please make it Python 2+3 compatible.

File: examples/mnist_dataset_api.py (hunk 5 above)
Comment (jsaporta) [reply examples/mnist_dataset_api.py]: Sorry, I find this comment a bit confusing. I use the `_train` arrays when I create the `dataset` object a few lines below. If I `_` them out, how will I get the training data to the model?

Would you prefer I have two separate calls to `mnist.load_data()`: one for the training data here and the other for the test data in the testing phase?

File: examples/mnist_dataset_api.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+train_model = keras.models.Model(inputs=model_input, outputs=model_output)
+
+train_model.compile(optimizer=keras.optimizers.RMSprop(lr=2e-3, decay=1e-5),
+                    loss='categorical_crossentropy',
+                    metrics=['accuracy'],
+                    target_tensors=[targets])
+train_model.summary()
+
+train_model.fit(epochs=epochs,
+                steps_per_epoch=steps_per_epoch)
```
Comment (lgeiger) [inline examples/mnist_dataset_api.py]: As a user I'm wondering if it is possible to use validation data together with this workflow?

Comment (fchollet) [reply examples/mnist_dataset_api.py]: On top of my mind, I think you can, but the model from draw validation samples/targets from its hard-coded queue, which in this example is the training data. Would have to check though.

Comment (lgeiger) [reply examples/mnist_dataset_api.py]: @fchollet Thanks for the fast answer.

For the record, I just naively tried adding `validation_data=(x_val, y_val)` to the `.fit` call, but it errors:
```bash
File "keras/engine/training.py", line 59, in _standardize_input_data
  'expected no data, but got:', data)
ValueError: ('Error when checking model input: expected no data, but got:', array([[[[ 0.],
...
```

Also using `validation_split` fails, but I guess that's expected:
```bash
File "keras/engine/training.py", line 1584, in fit
    if hasattr(x[0], 'shape'):
IndexError: list index out of range
```

**Pull request merged**